In [1]:
#import packages
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import requests
import time
import matplotlib as plt
%matplotlib inline

In [2]:
#Read csv file
df=pd.read_csv('twitter-archive-enhanced.csv')
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [3]:
df.shape

(2356, 17)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [5]:
#Exploring data
df.iloc[1]

tweet_id                                                     892177421306343426
in_reply_to_status_id                                                       NaN
in_reply_to_user_id                                                         NaN
timestamp                                             2017-08-01 00:17:27 +0000
source                        <a href="http://twitter.com/download/iphone" r...
text                          This is Tilly. She's just checking pup on you....
retweeted_status_id                                                         NaN
retweeted_status_user_id                                                    NaN
retweeted_status_timestamp                                                  NaN
expanded_urls                 https://twitter.com/dog_rates/status/892177421...
rating_numerator                                                             13
rating_denominator                                                           10
name                                    

In [6]:
df.iloc[1]['text']

"This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV"

In [7]:
df.iloc[1]['expanded_urls']

'https://twitter.com/dog_rates/status/892177421306343426/photo/1'

# Image Predictions

In [8]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [9]:
#The above output means it succesfully ran without any error, now we are going to open an empty file and write contents into it
with open('image_predictions.tsv', 'wb') as file:
    file.write(response.content)
# Read in tsv file as a Pandas DataFrame    
img = pd.read_csv('image_predictions.tsv', sep='\t')

In [10]:
img.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [11]:
img.shape

(2075, 12)

In [12]:
#Create a dictionary  to store twitter credentials
twitter_cred = dict()

#Enter your own consumer_key, consumer_secret, access_key and access_secret_key
consumer_key = 'XXXX'
consumer_secret ='XXXX'
access_token = 'XXXX'
access_secret = 'XXXX'

#Saving info to json so that it can be reused in code without exposing to public
with open('twitter_credentials.json', 'w') as secret_info:
 json.dump(twitter_cred, secret_info, indent=4, sort_keys=True)

In [13]:
#Create tweet API to connect to twitter with your credentials
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth_handler = auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)
# Create the api to connect to twitter with your credentials
api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), # Parse the result to Json Object
                 wait_on_rate_limit = True, # Automatically wait for rate limits to replenish
                 wait_on_rate_limit_notify = True) # Print a notification when Tweepy is waiting for rate limits to replenish

In [14]:
# List for storing result
df_list = []
#List for storing errors
error_list = []

# Calculate the time of excution
start = time.time()

# Get the tweet object for all the teweets in archive dataframe 
for tweet_id in df['tweet_id']:
    try:
        page = api.get_status(tweet_id, tweet_mode = 'extended')
        # Print one page to look at the structure of the returned file
        # and the names of attributes
        # print(json.dumps(page, indent = 4))
        #break
        
        favorites = page['favorite_count'] # How many favorites the tweet had
        retweets = page['retweet_count'] # Count of the retweet
        user_followers = page['user']['followers_count'] # How many followers the user had
        user_favourites = page['user']['favourites_count'] # How many favorites the user had
        date_time = page['created_at'] # The date and time of the creation
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
    
    # Catch the exceptions of the TweepError
    except Exception as e:
        print(str(tweet_id)+ " _ " + str(e))
        error_list.append(tweet_id)

# Calculate the time of execution
end = time.time()
print(end - start)

888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872261713294495745 _ [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 _ [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
856602993587888130 _ [{'code': 144, 'message': 'No status found with that ID.'}]
851953902622658560 _ [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 _ [{'code': 144, 'message': 'No status found with that ID.'}]
844704788403113984 _ [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
837366284874571778 _ [{'code

In [15]:
#Fetch the number results in df_list
len(df_list)

2332

In [16]:
#No. of errors
len(error_list)

24

It took us about 57 minutes approximately to fetch results

In [17]:
#Reiterating the process to fetch from error list
errors=[]
for e in error_list:
    try:
        favorites = page['favorite_count']
        retweets = page['retweet_count']
        user_followers = page['user']['followers_count']
        user_favourites = page['user']['favourites_count']
        date_time = page['created_at']
        
        df_list.append({'tweet_id': int(tweet_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets),
                        'user_followers': int(user_followers),
                        'user_favourites': int(user_favourites),
                        'date_time': pd.to_datetime(date_time)})
        
    except Exception:
        #In case of erros, add the tweet id to errors 
        print(str(tweet_id)+ " _ " + str(e))
        errors.append(e)

In [18]:
#Confirming the no. of errors
len(errors)

0

In [19]:
#Create a dataframe
tweets_json=pd.DataFrame(df_list,columns=['tweet_id','favorites','retweets','user_followers','user_favourites','date_time'])

In [20]:
# Save the dataframe to json.txt 
tweets_json.to_csv('tweet_json.txt', encoding = 'utf-8', index=False)

In [21]:
#Read json.txt into dataframe
tweet=pd.read_csv('tweet_json.txt',encoding='utf-8')
tweet.head()

tweet_id  favorites  retweets  user_followers  user_favourites  \
0  892420643555336193      37200      7969         8309109           142727   
1  892177421306343426      31994      5913         8309109           142727   
2  891815181378084864      24097      3909         8309109           142727   
3  891689557279858688      40512      8110         8309109           142727   
4  891327558926688256      38749      8803         8309109           142727   

                   date_time  
0  2017-08-01 16:23:56+00:00  
1  2017-08-01 00:17:27+00:00  
2  2017-07-31 00:18:03+00:00  
3  2017-07-30 15:58:51+00:00  
4  2017-07-29 16:00:24+00:00

In [22]:
tweet.shape

(2356, 6)

In [23]:
tweet['tweet_id'].duplicated().value_counts()

False    2332
True       24
Name: tweet_id, dtype: int64

In [24]:
tweet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 6 columns):
tweet_id           2356 non-null int64
favorites          2356 non-null int64
retweets           2356 non-null int64
user_followers     2356 non-null int64
user_favourites    2356 non-null int64
date_time          2356 non-null object
dtypes: int64(5), object(1)
memory usage: 110.5+ KB


In [25]:
tweet.to_csv("outputfile.csv", sep=",", decimal=",")

<h4>Gathering Conclusions:</h4>
    1. Gathered data from various sources and saved them into a dataframe for next steps <br>
    2. 'twitter_archive_enhanced.csv' has been given to us. It has been read into <b> df</b> dataframe and consists of 2356 rows.<br>
    3. The tweet image prediction file (.tsv) has been downloaded programatically using requests library from the url below:
     https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv <br>
    4. Using tweet_id, scraping Twitter APi using tweepy and storing each tweet's entire set of json data into <b>tweet_json.txt</b>.This dataframe consists of around 2356 records with 24 duplicate tweet ids.
    
    